kernel/arch/dreamcast/include/dc/cdrom.h

```diff
@@ -187,8 +189,18 @@ __BEGIN_DECLS
     cdrom_read_sectors_ex.
     @{
 */
-#define CDROM_READ_PIO 0    /**< \brief Read sector(s) in PIO mode */
-#define CDROM_READ_DMA 1    /**< \brief Read sector(s) in DMA mode */
+/**< \brief Read sector(s) in PIO mode with polling syscalls in current thread */
+#define CDROM_READ_PIO 0
+/**< \brief Read sector(s) in DMA mode with polling syscalls in current thread */
+#define CDROM_READ_DMA 1
+/**< \brief Read sector(s) in DMA mode, but instead of polling syscalls
+ * and passing the current thread, a vblank and DMA IRQs with a semaphores are used.
+ */
+#define CDROM_READ_DMA_IRQ 2
+/**< \brief Read sector(s) in PIO mode, but instead of polling syscalls
+ * and passing the current thread, a vblank IRQ with a semaphore are used.
+ */
+#define CDROM_READ_PIO_IRQ 3
```

Comment:
What's the advantage of using `CDROM_READ_PIO` vs `CDROM_READ_PIO_IRQ`, or `CDROM_READ_DMA` vs. `CDROM_READ_DMA_IRQ`?

From what I understand the _IRQ modes will always be better, so they should just replace the old ones.

Reply:
There are several reasons why I left the old modes.
1) Backward compatibility.
2) If you use one thread in the program, which also waits for data from the drive, then it is better not to use IRQs at all, this will work a little faster.
3) If I do all the code on interrupts, then I will break the support of serial SD card in the DreamShell ISO Loader. Therefore, in ISO FS I left a fallback, if GD streams return an error (SD card will do this), then I read the data in the regular sector mode without interrupts. I haven't figured out how to emulate interrupts yet, so for now I decided to do it this way. I wouldn't want to deprive SD card users of new homebrews :)

Reply:
1. Well then we just need to make the new ones backwards-compatible :)
2. Did you benchmark that? I don't see how waiting for an IRQ would be slower than constantly yielding the thread.
3. You introduced two mechanisms to detect the end of transfer: the "DMA done" IRQ, and a polling done in the VBLANK IRQ. You could slightly modify the VBLANK handler to also handle the "DMA done" use-case, so that it would still work when the hardware never sends you the DMA IRQ.

Reply:
2. Yes, and it was slower. Because vblank does not happen as often as I would like for this case. Especially if it concerns fast devices like flash. But for DreamShell I made a workaround here, before going into waiting for a command result from an vblank, I make one pass of the server exec and check the command status. In DreamShell, pre-reading is done instantly and therefore I do not go into waiting for the result of vblank work. But if I remove these two lines, I will get a slowdown that is even noticeable by eye:
https://github.com/KallistiOS/KallistiOS/pull/736/files#diff-76ed3876780fab6803fad0fa9fb1378cf3871f20bb9704d3501c9429f2d3618eR154
3. Here everything is the same and in #2. The interrupt from DMA will arrive earlier than the vblank will work. All this is noticeable when there are a lot of small readings.
This vblank usage only good for slow laser head moving.